src/pip/_internal/commands/install.py

```diff
@@ -732,3 +737,25 @@ def create_os_error_message(error, show_traceback, using_user_site):
         parts.append(".\n")
 
     return "".join(parts).strip() + "\n"
+
+
+def prepare_sudo_warning():
+    # type: () -> None
+    """Output a warning for sudo users on Unix.
+
+    On Windows, users may run pip as Administrator without issues.
+    """
+    try:
+        if sys.platform == 'darwin' or sys.platform == 'linux':
+            warn_sudo = os.getuid() == 0
+        elif sys.platform == 'win32' or sys.platform == 'cygwin':
+            warn_sudo = False
+    except AttributeError:
+        warn_sudo = False
+
+    if warn_sudo and not running_under_virtualenv():
```

Comment:
I'd strongly prefer using early returns, instead of the current structure involving boolean and multiple combinational conditions.

I'd also like to use the `WINDOWS` constant from pip's compat module. 

```py
def warn_if_run_as_root():
    if running_under_virtualenv():
        return
    if WINDOWS:
        return
    if not hasattr(os, "getuid"):
        return
    if os.getuid() != 0:
        return
    logger.warning("...")
```

Reply:
I like this format. But the reason for the current (admittedly weird) if/elif structure is for [Mypy platform checking](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks).

Your example would fail Mypy since `if os.getuid() != 0:` will still be run on Mypy's Windows linter, throwing an AttributeError and failing the test suite.

Reply:
The weird if/elif structure was to follow [Mypy's platform check](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks) convention and avoid AttributeErrors with `os.getuid`.

But I like this version better.

Reply:
Update: this still fails Mypy Windows despite the `hasattr` :/

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src\pip\_internal\cli\req_command.py:177: error: Module has no attribute
"getuid"; maybe "getpid" or "getppid"?
        if os.getuid() != 0:
           ^
Found 1 error in 1 file (checked 140 source files)
```

Reverting to the previous version.

Reply:
Mypy-compatible version:
```
    if running_under_virtualenv():                                                                  
        return                                                                                      
    if not hasattr(os, "getuid"):                                                                   
        return                                                                                      
    if sys.platform == "win32" or sys.platform == "cygwin":                                         
        return                                                                                      
    elif sys.platform == "darwin" or sys.platform == "linux":                                       
        if os.getuid() != 0:                                                                        
            return                                                                                  
    logger.info("...")
```